resolwe/storage/cleanup.py

```diff
@@ -0,0 +1,78 @@
+"""Storage cleanup."""
+import logging
+from typing import Optional
+
+from django.db import transaction
+
+from resolwe.storage.models import FileStorage, StorageLocation
+
+logger = logging.getLogger(__name__)
+
+
+class Cleaner:
+    """Remove unreferenced data.
+
+    Remove data from StorageLocation objects that are no longer referenced
+    by any Data object.
+    """
+
+    def _cleanup(self, storage_location: StorageLocation):
+        """Delete data from StorageLocation object."""
+        # Make sure this will get writen to the database.
+        # Problem is we are inside transaction so what happens if the
+        # connection to the database fails or database crashed or...
+        connector = storage_location.connector
+        if connector is None:
+            logger.error(
+                "Unable to cleanup StorageLocation {}: connector not found.".format(
+                    storage_location.id
+                )
+            )
+            return
+        try:
+            connector.delete(storage_location.urls)
+        except Exception:
+            logger.exception(
+                "Exception while deleting StorageLocation {}.".format(
+                    storage_location.id
+                )
+            )
+        # Delete the storage location when all files have been removed.
+        storage_location.delete()
+
+    def _process_file_storage(self, file_storage: FileStorage):
+        """Delete all data from FileStorage object."""
+        # Do not remove locked StorageLocation.
+        for storage_location in file_storage.storage_locations.filter(
+            status=StorageLocation.STATUS_DELETING
+        ):
+            self._cleanup(storage_location)
+
+        if file_storage.storage_locations.count() == 0:
+            file_storage.delete()
+
+    def process(self, file_storage_id: Optional[int] = None):
+        """Process objects to clean.
+
+        When file_storage is not None process only that object.
+        """
+        logger.debug("Starting cleanup manager run")
+        # Set applicable storage locations to deleting.
+        StorageLocation.all_objects.unreferenced_locations().update(
```

Comment:
All locations are updated, even if `file_storage_id` is set. It would be better if you move this inside the for loop, just before the transaction.